### ISSUE TITLE
registerProducer and unregisterProducer need special treatment in twisted.internet._newtls

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5347]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5347|
|Type|defect|
|Created|2011-10-31 13:24:36Z|
|Branch|https://github.com/twisted/twisted/tree/producer-tls-5347-2|

As with methods like `write()`, `registerProducer` and `unregisterProducer` need to be forwarded to the `TLSMemBIOProtocol` after `startTLS` is called, while the TLS protocol still needs access to the underlying transport's implementation.

Apparently we need better producer tests in order to catch this... or perhaps I'm just confused.

<details><summary>Searchable metadata</summary>

```
trac-id__5347 5347
type__defect defect
reporter__itamar itamar
priority__high high
milestone__ 
branch__branches_producer_tls_5347_2 branches/producer-tls-5347-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1320067476000000 1320067476000000
changetime__1323374336000000 1323374336000000
version__None None
owner__itamar itamar
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [33033]) Branching to 'producer-tls-5347'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

It'd be good to merge this into 11.1... but so far I'm getting mysterious test failures on Windows.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/producer-tls-5347

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

OK, better now. Ready for review. Tests pass, other than random test errors on Windows that don't always happen. (At a guess some these are due to connecting to listeningPort.getHost().host, instead of "127.0.0.1", given past issues we've had with that and the fact that fixing it in my new test made it consistently pass).

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/producer-tls-5347

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks!

  1. If the goal of `def registerProducer(self, *args, **kwargs):` in `_BypassTLS` is to avoid having to know the signature of `registerProducer`, it seems defeated by the two other new implementations of that method that are spelled `registerProducer(self, producer, streaming):`
  1. The case where `startTLS` comes between a `registerProducer` call and the corresponding `unregisterProducer` call seems like it will break now.
  1. The test has some issues:
    1. I would really like it if we stopped using the global reactor in tests.  Elaboration at <http://twistedmatrix.com/pipermail/twisted-python/2011-November/024732.html>
    1. The test mixes together testing of TCP/startTLS and "native" TLS.  It's really at least two different tests.
    1. ''TLSMemoryBioProtocol'' is miscapitalized in all the comments

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

Removing milestone.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

So looking at _newtls, there's an additional issue of the buffering layer that it provides post-startTLS, which presumably should also interact with producers. However, given the existence of the buffering during handshakes in TLSMemoryBIOProtocol, why do we need this extra buffering?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

`_oldtls` needs it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

(In [33056]) Branching to 'producer-tls-5347-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst removed owner|
|-|-|

Review comments have been addressed, and I removed the unnecessary extra buffering layer in `_newtls`, which as we concluded on IRC is not needed when using `TLSMemoryBIOProtocol`.

The first time I ran the windows tests, lots of seemingly unrelated tests failed (e.g. xmlrpc). Re-running seemed to fix it, so I'm guessing it was a load issue -- one of the bots ran twice as long as usual:

buildbot.twistedmatrix.com/boxes-supported?branch=/branches/producer-tls-5347-2

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @itamarst|
|-|-|

It looks good. The tests are somewhat terrible (the `_producer` access for example), but I know the pain. Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set status to `closed`|
|-|-|

(In [33299]) Merge producer-tls-5347-2.

Fixes: [#5347](https://github.com/twisted/twisted/issues/5347)
Review: exarkun, therve
Author: itamar

Fix rare situation where producers wouldn't be paused when using transport.startTLS.
